Subject: [PATCH] STE U8500 video recording fix

U8500 video codecs require storing metadata before entering the idle state.
Port kRequiresStoreMetaDataBeforeIdle quirk to ACodec as originally implemented
in OMXCodec.

diff --git a/media/libstagefright/ACodec.cpp b/media/libstagefright/ACodec.cpp
index 72d340b91..43959ed88 100755
--- a/media/libstagefright/ACodec.cpp
+++ b/media/libstagefright/ACodec.cpp
@@ -1872,6 +1872,7 @@ status_t ACodec::configureCodec(
     }
 
     int32_t storeMeta;
+#ifndef STE_HARDWARE
     if (encoder
             && msg->findInt32("android._input-metadata-buffer-type", &storeMeta)
             && storeMeta != kMetadataBufferTypeInvalid) {
@@ -1905,6 +1906,7 @@ status_t ACodec::configureCodec(
                     "using-sw-read-often", !!(usageBits & GRALLOC_USAGE_SW_READ_OFTEN));
         }
     }
+#endif
 
     int32_t prependSPSPPS = 0;
     if (encoder
@@ -1943,12 +1945,14 @@ status_t ACodec::configureCodec(
             && msg->findInt32("android._store-metadata-in-buffers-output", &storeMeta)
             && storeMeta != 0);
 
+#ifndef STE_HARDWARE
         mOutputMetadataType = kMetadataBufferTypeNativeHandleSource;
         err = mOMX->storeMetaDataInBuffers(mNode, kPortIndexOutput, enable, &mOutputMetadataType);
         if (err != OK) {
             ALOGE("[%s] storeMetaDataInBuffers (output) failed w/ err %d",
                 mComponentName.c_str(), err);
         }
+#endif
 
         if (!msg->findInt64(
                     "repeat-previous-frame-after",
@@ -6917,6 +6921,17 @@ ACodec::LoadedState::LoadedState(ACodec *codec)
 void ACodec::LoadedState::stateEntered() {
     ALOGV("[%s] Now Loaded", mCodec->mComponentName.c_str());
 
+#ifdef STE_HARDWARE
+    int err;
+
+    if (mCodec->mQuirks & kRequiresStoreMetaDataBeforeIdle) {
+        ALOGV("%s: chrono: storing metadata before idle", mCodec->mComponentName.c_str());
+        err = mCodec->mOMX->storeMetaDataInBuffers(mCodec->mNode, kPortIndexInput, OMX_TRUE);
+        if (err != OK)
+            ALOGE("Storing meta data in video buffers is not supported");
+    }
+#endif
+
     mCodec->mPortEOS[kPortIndexInput] =
         mCodec->mPortEOS[kPortIndexOutput] = false;
 
diff --git a/media/libstagefright/ACodec.cpp b/media/libstagefright/ACodec.cpp
index d0f8ada00..d94cb2998 100644
--- a/media/libstagefright/ACodec.cpp
+++ b/media/libstagefright/ACodec.cpp
@@ -1776,6 +1776,10 @@ void OMXCodec::setComponentRole(
             "video_decoder.mpeg4", "video_encoder.mpeg4" },
         { MEDIA_MIMETYPE_VIDEO_H263,
             "video_decoder.h263", "video_encoder.h263" },
+#ifdef STE_HARDWARE
+        { MEDIA_MIMETYPE_VIDEO_VC1,
+            "video_decoder.vc1", "video_encoder.vc1" },
+#endif
         { MEDIA_MIMETYPE_VIDEO_VP8,
             "video_decoder.vp8", "video_encoder.vp8" },
         { MEDIA_MIMETYPE_VIDEO_VP9,
